codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/OperationGenerator.java

```diff
@@ -116,6 +131,30 @@ public void run() {
             writer.writeDocs("Metadata pertaining to the operation's result.");
             writer.write("ResultMetadata $T", metadataSymbol);
         });
+
+        if (idempotencyTokenMemberShape != null) {
+            // Generate Middleware for Idempotency token trait
+            GoStackStepMiddlewareGenerator middleware =
+                    GoStackStepMiddlewareGenerator.createInitializeStepMiddleware(
+                            getIdempotencyTokenMiddlewareName(operation));
+
+            String memberName = symbolProvider.toMemberName(idempotencyTokenMemberShape);
+            middleware.writeMiddleware(writer, (generator, middlewareWriter) -> {
+                middlewareWriter.write("input, ok := in.Parameters.($P)", inputSymbol);
+                middlewareWriter.write("if !ok { return out, metadata, "
+                        + "fmt.Errorf(\"expected middleware input to be of type $P \")}", inputSymbol);
+
+                middlewareWriter.openBlock("if input.$L == nil {", "}", memberName, () -> {
+                    String operand = "v";
+                    integrations.forEach(goIntegration -> {
+                        goIntegration.generateValueForIdempotencyToken(middlewareWriter, operand);
+                    });
+
+                    middlewareWriter.write("input.$L = &$L", memberName, operand);
+                });
+                middlewareWriter.write("return next.HandleInitialize(ctx, in)");
```

Comment:
nit: The generator given to your lambda can give you the name of the handler to call. So this could be `next.$L(ctx, in)` and pass in `generator.getHandleMethodName()` for the `$L` value.